File: src/test/java/com/mikea/bayes/examples/ProgrammingAssignment2.java
package com.mikea.bayes.examples;

import com.mikea.bayes.Factor;
import com.mikea.bayes.Var;
import com.mikea.bayes.VarAssignment;
import org.junit.Test;

import static com.google.common.base.Preconditions.checkArgument;
import static com.mikea.bayes.Utils.strings;
import static com.mikea.bayes.VarAssignment.at;
import static org.junit.Assert.assertEquals;

/**
 *
 * Programming assignment 2 for PGM.
 *
 * @author mike.aizatsky@gmail.com
 */
public class ProgrammingAssignment2 {
    @Test
    public void testPhenotypeGivenGenotypeMendelianFactor() throws Exception {
        // 0 - FF
        // 1 - Ff
        // 2 - ff
        Var genotype = new Var("genotype", 3, strings("FF", "Ff", "ff"));
        Var phenotype = new Var("phenotype", 2, strings("F", "T"));

        assertEquals(
                "Factor({phenotype(2, [F, T]), genotype(3, [FF, Ff, ff])}):\n" +
                        "{genotype=FF}: 0.0 1.0\n" +
                        "{genotype=Ff}: 0.0 1.0\n" +
                        "{genotype=ff}: 1.0 0.0\n",
                getPhenotypeGivenGenotypeMendelianFactor(true, genotype, phenotype).toStringAsTable(phenotype, "%.1f"));

        assertEquals(
                "Factor({phenotype(2, [F, T]), genotype(3, [FF, Ff, ff])}):\n" +
                        "{genotype=FF}: 1.0 0.0\n" +
                        "{genotype=Ff}: 1.0 0.0\n" +
                        "{genotype=ff}: 0.0 1.0\n",
                getPhenotypeGivenGenotypeMendelianFactor(false, genotype, phenotype).toStringAsTable(phenotype, "%.1f"));
    }

    @Test
    public void testPhenotypeGivenGenotypeFactor() throws Exception {
        Var genotype = new Var("genotype", 3);
        Var phenotype = new Var("phenotype", 2);

        assertEquals("Factor({phenotype(2), genotype(3)}):\n" +
                "{genotype=0}: 0.2 0.8\n" +
                "{genotype=1}: 0.4 0.6\n" +
                "{genotype=2}: 0.9 0.1\n",
                getPhenotypeGivenGenotypeFactor(genotype, phenotype, new double[] {0.8, 0.6, 0.1}).toStringAsTable(phenotype, "%.1f"));
    }

    @Test
    public void testGenotypeGivenAlleleFreqsFactor() throws Exception {
        Var genotype = new Var("genotype", 3);
        assertEquals("Factor({genotype(3)}, [0.010000000000000002, 0.18000000000000002, 0.81])",
                getGenotypeGivenAlleleFreqsFactor(genotype, new double[]{0.1, 0.9}).toString());
    }

    @Test
    public void testGenotypeGivenParentsGenotypesFactor() throws Exception {
        Var genotypeParent1 = new Var("p1", 3);
        Var genotypeParent2 = new Var("p2", 3);
        Var genotypeChild = new Var("c", 3);

        assertEquals(
                "Factor({c(3), p1(3), p2(3)}):\n" +
                        "{p1=0, p2=0}: 1.00 0.00 0.00\n" +
                        "{p1=0, p2=1}: 0.50 0.50 0.00\n" +
                        "{p1=0, p2=2}: 0.00 1.00 0.00\n" +
                        "{p1=1, p2=0}: 0.50 0.50 0.00\n" +
                        "{p1=1, p2=1}: 0.25 0.50 0.25\n" +
                        "{p1=1, p2=2}: 0.00 0.50 0.50\n" +
                        "{p1=2, p2=0}: 0.00 1.00 0.00\n" +
                        "{p1=2, p2=1}: 0.00 0.50 0.50\n" +
                        "{p1=2, p2=2}: 0.00 0.00 1.00\n",
                getGenotypeGivenParentsGenotypesFactor(genotypeChild, genotypeParent1, genotypeParent2, 2).toStringAsTable(genotypeChild, "%1.2f"));
    }

    @Test
    public void testGenotypeAllelesMap() throws Exception {
        int numAlleles = 5;
        int numGenotypes = getNumGenotypes(numAlleles);
        int[][] allelesToGenotypeMap = getAllelesToGenotypeMap(numAlleles);
        int[][] genotypeToAllelesMap = getGenotypeToAllelesMap(numAlleles);


        for (int i = 0; i < numGenotypes; ++i) {
            assertEquals(i, allelesToGenotypeMap[genotypeToAllelesMap[i][0]][genotypeToAllelesMap[i][1]]);
        }

        for (int i = 0; i < numAlleles; ++i) {
            for (int j = i; j < numAlleles; ++j) {
                int idx = allelesToGenotypeMap[i][j];
                assertEquals(genotypeToAllelesMap[idx][0], i);
                assertEquals(genotypeToAllelesMap[idx][1], j);
            }
        }

    }

    private Factor getGenotypeGivenParentsGenotypesFactor(Var genotypeChild, Var genotypeParent1, Var genotypeParent2, int numAlleles) {
        int numGenotypes = getNumGenotypes(numAlleles);

        checkArgument(genotypeChild.getCardinality() == numGenotypes);
        checkArgument(genotypeParent1.getCardinality() == numGenotypes);
        checkArgument(genotypeParent2.getCardinality() == numGenotypes);
        checkArgument(genotypeChild != genotypeParent1);
        checkArgument(genotypeChild != genotypeParent2);
        checkArgument(genotypeParent1 != genotypeParent2);

        Factor.Builder builder = Factor.withVariables(genotypeChild, genotypeParent1, genotypeParent2);


        int[][] allelesToGenotypeMap = getAllelesToGenotypeMap(numAlleles);
        int[][] genotypeToAllelesMap = getGenotypeToAllelesMap(numAlleles);


        for (int genotype1 = 0; genotype1 < numGenotypes; ++genotype1) {
            int allele11 = genotypeToAllelesMap[genotype1][0];
            int allele12 = genotypeToAllelesMap[genotype1][1];

            for (int genotype2 = 0; genotype2 < numGenotypes; ++genotype2) {
                int allele21 = genotypeToAllelesMap[genotype2][0];
                int allele22 = genotypeToAllelesMap[genotype2][1];

                int childGenotype1 = allelesToGenotypeMap[allele11][allele21];
                int childGenotype2 = allelesToGenotypeMap[allele12][allele21];
                int childGenotype3 = allelesToGenotypeMap[allele11][allele22];
                int childGenotype4 = allelesToGenotypeMap[allele12][allele22];

                builder = builder.add(
                        at(genotypeParent1, genotype1).at(genotypeParent2, genotype2).at(genotypeChild, childGenotype1),
                        1
                );
                builder = builder.add(
                        at(genotypeParent1, genotype1).at(genotypeParent2, genotype2).at(genotypeChild, childGenotype2),
                        1
                );
                builder = builder.add(
                        at(genotypeParent1, genotype1).at(genotypeParent2, genotype2).at(genotypeChild, childGenotype3),
                        1
                );
                builder = builder.add(
                        at(genotypeParent1, genotype1).at(genotypeParent2, genotype2).at(genotypeChild, childGenotype4),
                        1
                );
            }
        }

        return builder.build().normalizeBy(genotypeParent1, genotypeParent2);
    }

    private int getNumGenotypes(int numAlleles) {
        return numAlleles * (numAlleles - 1) / 2 + numAlleles;
    }

    private Factor getGenotypeGivenAlleleFreqsFactor(Var genotype, double[] alleleFreqs) {
        int numAlleles = alleleFreqs.length;
        int numGenotypes = getNumGenotypes(numAlleles);

        int[][] allelesToGenotypeMap = getAllelesToGenotypeMap(numAlleles);

        double[] values = new double[numGenotypes];
        for (int i = 0; i < numAlleles; ++i) {
            for (int j = i; j < numAlleles; ++j) {
                values[allelesToGenotypeMap[i][j]] = alleleFreqs[i] * alleleFreqs[j] * (i == j ? 1 : 2);
            }
        }

        return Factor
                .withVariables(genotype)
                .row(genotype, new VarAssignment.Builder(), values)  // todo: this new VarAssignment.Builder() doesn't look good.
                .build();
    }

    private int[][] getAllelesToGenotypeMap(int numAlleles) {
        int[][] map = new int[numAlleles][];

        for (int i = 0; i < numAlleles; ++i) {
            map[i] = new int[numAlleles];
        }

        for (int idx = 0, i = 0; i < numAlleles; ++i) {
            for (int j = i; j < numAlleles; ++j, ++idx) {
                map[i][j] = idx;
                map[j][i] = idx;
            }
        }

        return map;
    }

    private int[][] getGenotypeToAllelesMap(int numAlleles) {
        int numGenotypes = getNumGenotypes(numAlleles);
        int[][] map = new int[numGenotypes][];

        for (int idx = 0, i = 0; i < numAlleles; ++i) {
            for (int j = i; j < numAlleles; ++j, ++idx) {
                map[idx] = new int[2];
                map[idx][0] = i;
                map[idx][1] = j;
            }
        }

        return map;
    }

    private Factor getPhenotypeGivenGenotypeFactor(Var genotype, Var phenotype, double[] alphas) {
        checkArgument(phenotype.getCardinality() == 2);
        checkArgument(alphas.length == genotype.getCardinality());

        double[] alphasCompliments = new double[alphas.length];
        for (int i = 0; i < alphas.length; i++) {
            alphasCompliments[i] = 1 - alphas[i];
        }

        return Factor
                .withVariables(phenotype, genotype)
                .row(genotype, at(phenotype, 0), alphasCompliments)
                .row(genotype, at(phenotype, 1), alphas)
                .build();
    }

    private Factor getPhenotypeGivenGenotypeMendelianFactor(boolean isDominant,
                                                            Var genotype,
                                                            Var phenotype) {
        checkArgument(genotype.getCardinality() == 3);
        checkArgument(phenotype.getCardinality() == 2);

        if (isDominant) {
            return Factor
                    .withVariables(phenotype, genotype)
                    .row(phenotype, at(genotype, "FF"), strings("T", "F"), new double[]{1, 0})
                    .row(phenotype, at(genotype, "Ff"), strings("T", "F"), new double[]{1, 0})
                    .row(phenotype, at(genotype, "ff"), strings("T", "F"), new double[]{0, 1})
                    .build();
        } else {
            return Factor
                    .withVariables(phenotype, genotype)
                    .row(phenotype, at(genotype, "FF"), strings("T", "F"), new double[]{0, 1})
                    .row(phenotype, at(genotype, "Ff"), strings("T", "F"), new double[]{0, 1})
                    .row(phenotype, at(genotype, "ff"), strings("T", "F"), new double[]{1, 0})
                    .build();
        }
    }

}

File: src/test/java/com/mikea/bayes/examples/ProgrammingAssignment2.scala
package com.mikea.bayes.examples

import org.junit.Test
import org.junit.Assert.assertEquals
import com.mikea.bayes.{VarAssignment, Factor, ProbabilitySpace, Var}
import com.mikea.bayes.VarAssignment.at
import com.mikea.bayes.Factor.Builder

/**
 *
 * Programming assignment 2 for PGM.
 *
 * @author mike.aizatsky@gmail.com
 */
class ProgrammingAssignment2 {
  @Test def testPhenotypeGivenGenotypeMendelianFactor() {
    val space: ProbabilitySpace = new ProbabilitySpace
    val genotype: Var = space.newVar("genotype", 3, List("FF", "Ff", "ff"))
    val phenotype: Var = space.newVar("phenotype", 2, List("F", "T"))
    assertEquals("Factor({phenotype, genotype}):\n" + "{genotype=FF}: 0.0 1.0\n" + "{genotype=Ff}: 0.0 1.0\n" + "{genotype=ff}: 1.0 0.0\n", getPhenotypeGivenGenotypeMendelianFactor(isDominant = true, genotype, phenotype).toStringAsTable(phenotype, "%.1f"))
    assertEquals("Factor({phenotype, genotype}):\n" + "{genotype=FF}: 1.0 0.0\n" + "{genotype=Ff}: 1.0 0.0\n" + "{genotype=ff}: 0.0 1.0\n", getPhenotypeGivenGenotypeMendelianFactor(isDominant = false, genotype, phenotype).toStringAsTable(phenotype, "%.1f"))
  }

  @Test def testPhenotypeGivenGenotypeFactor() {
    val space: ProbabilitySpace = new ProbabilitySpace
    val genotype: Var = space.newVar("genotype", 3)
    val phenotype: Var = space.newVar("phenotype", 2)
    assertEquals("Factor({phenotype, genotype}):\n" + "{genotype=0}: 0.2 0.8\n" + "{genotype=1}: 0.4 0.6\n" + "{genotype=2}: 0.9 0.1\n", getPhenotypeGivenGenotypeFactor(genotype, phenotype, Array[Double](0.8, 0.6, 0.1)).toStringAsTable(phenotype, "%.1f"))
  }

  @Test def testGenotypeGivenAlleleFreqsFactor() {
    val space: ProbabilitySpace = new ProbabilitySpace
    val genotype: Var = space.newVar("genotype", 3)
    assertEquals("Factor({genotype}, [0.010000000000000002, 0.18000000000000002, 0.81])", getGenotypeGivenAlleleFreqsFactor(genotype, Array[Double](0.1, 0.9)).toString)
  }

  @Test def testGenotypeGivenParentsGenotypesFactor() {
    val space: ProbabilitySpace = new ProbabilitySpace
    val genotypeParent1: Var = space.newVar("p1", 3)
    val genotypeParent2: Var = space.newVar("p2", 3)
    val genotypeChild: Var = space.newVar("c", 3)
    assertEquals("Factor({c, p1, p2}):\n" + "{p1=0, p2=0}: 1.00 0.00 0.00\n" + "{p1=0, p2=1}: 0.50 0.50 0.00\n" + "{p1=0, p2=2}: 0.00 1.00 0.00\n" + "{p1=1, p2=0}: 0.50 0.50 0.00\n" + "{p1=1, p2=1}: 0.25 0.50 0.25\n" + "{p1=1, p2=2}: 0.00 0.50 0.50\n" + "{p1=2, p2=0}: 0.00 1.00 0.00\n" + "{p1=2, p2=1}: 0.00 0.50 0.50\n" + "{p1=2, p2=2}: 0.00 0.00 1.00\n", getGenotypeGivenParentsGenotypesFactor(genotypeChild, genotypeParent1, genotypeParent2, 2).toStringAsTable(genotypeChild, "%1.2f"))
  }

  @Test def testGenotypeAllelesMap() {
    val numAlleles: Int = 5
    val numGenotypes: Int = getNumGenotypes(numAlleles)
    val allelesToGenotypeMap: Array[Array[Int]] = getAllelesToGenotypeMap(numAlleles)
    val genotypeToAllelesMap: Array[Array[Int]] = getGenotypeToAllelesMap(numAlleles)
    for (i <- 0 until numGenotypes) {
      assertEquals(i, allelesToGenotypeMap(genotypeToAllelesMap(i)(0))(genotypeToAllelesMap(i)(1)))
    }
    for (i <- 0 until numAlleles) {
      for (j <- i until numAlleles) {
        val idx: Int = allelesToGenotypeMap(i)(j)
        assertEquals(genotypeToAllelesMap(idx)(0), i)
        assertEquals(genotypeToAllelesMap(idx)(1), j)
      }
    }
  }

  private def getGenotypeGivenParentsGenotypesFactor(genotypeChild: Var, genotypeParent1: Var, genotypeParent2: Var, numAlleles: Int): Factor = {
    val numGenotypes: Int = getNumGenotypes(numAlleles)
    assert(genotypeChild.cardinality == numGenotypes)
    assert(genotypeParent1.cardinality == numGenotypes)
    assert(genotypeParent2.cardinality == numGenotypes)
    assert(genotypeChild ne genotypeParent1)
    assert(genotypeChild ne genotypeParent2)
    assert(genotypeParent1 ne genotypeParent2)
    var builder: Builder = Factor.withVariables(genotypeChild, genotypeParent1, genotypeParent2)
    val allelesToGenotypeMap: Array[Array[Int]] = getAllelesToGenotypeMap(numAlleles)
    val genotypeToAllelesMap: Array[Array[Int]] = getGenotypeToAllelesMap(numAlleles)
    for (genotype1 <- 0 until numGenotypes) {
      val allele11: Int = genotypeToAllelesMap(genotype1)(0)
      val allele12: Int = genotypeToAllelesMap(genotype1)(1)
      for (genotype2 <- 0 until numGenotypes) {
        val allele21: Int = genotypeToAllelesMap(genotype2)(0)
        val allele22: Int = genotypeToAllelesMap(genotype2)(1)
        val childGenotype1: Int = allelesToGenotypeMap(allele11)(allele21)
        val childGenotype2: Int = allelesToGenotypeMap(allele12)(allele21)
        val childGenotype3: Int = allelesToGenotypeMap(allele11)(allele22)
        val childGenotype4: Int = allelesToGenotypeMap(allele12)(allele22)
        builder = builder.add(at(genotypeParent1, genotype1).at(genotypeParent2, genotype2).at(genotypeChild, childGenotype1), 1)
        builder = builder.add(at(genotypeParent1, genotype1).at(genotypeParent2, genotype2).at(genotypeChild, childGenotype2), 1)
        builder = builder.add(at(genotypeParent1, genotype1).at(genotypeParent2, genotype2).at(genotypeChild, childGenotype3), 1)
        builder = builder.add(at(genotypeParent1, genotype1).at(genotypeParent2, genotype2).at(genotypeChild, childGenotype4), 1)
      }
    }

    builder.build().normalizeBy(genotypeParent1, genotypeParent2)
  }

  private def getNumGenotypes(numAlleles: Int): Int = { numAlleles * (numAlleles - 1) / 2 + numAlleles }

  private def getGenotypeGivenAlleleFreqsFactor(genotype: Var, alleleFreqs: Array[Double]): Factor = {
    val numAlleles: Int = alleleFreqs.length
    val numGenotypes: Int = getNumGenotypes(numAlleles)
    val allelesToGenotypeMap: Array[Array[Int]] = getAllelesToGenotypeMap(numAlleles)
    val values: Array[Double] = new Array[Double](numGenotypes)
    for (i <- 0 until numAlleles) {
      for (j <- 0 until numAlleles) {
        values(allelesToGenotypeMap(i)(j)) = alleleFreqs(i) * alleleFreqs(j) * (if (i == j) 1 else 2)
      }
    }
    Factor.withVariables(genotype).row(genotype, new VarAssignment.Builder(), values).build()
  }

  private def getAllelesToGenotypeMap(numAlleles: Int): Array[Array[Int]] = {
    val map: Array[Array[Int]] = new Array[Array[Int]](numAlleles)
    for (i <- 0 until numAlleles) {
      map(i) = new Array[Int](numAlleles)
    }

    var idx: Int = 0

    for (i <- 0 until numAlleles) {
      for (j <- i until numAlleles) {
        map(i)(j) = idx
        map(j)(i) = idx
        idx += 1
      }
    }

    map
  }

  private def getGenotypeToAllelesMap(numAlleles: Int): Array[Array[Int]] = {
    val numGenotypes: Int = getNumGenotypes(numAlleles)
    val map: Array[Array[Int]] = new Array[Array[Int]](numGenotypes)
    var idx: Int = 0

    for (i <- 0 until numAlleles) {
      for (j <- i until numAlleles) {
                map(idx) = new Array[Int](2)
                map(idx)(0) = i
                map(idx)(1) = j
        idx += 1
      }
    }

    map
  }

  private def getPhenotypeGivenGenotypeFactor(genotype: Var, phenotype: Var, alphas: Array[Double]): Factor = {
    assert(phenotype.cardinality == 2)
    assert(alphas.length == genotype.cardinality)
    val alphasCompliments: Array[Double] = new Array[Double](alphas.length)

    for (i <- 0 until alphas.length) {
          alphasCompliments(i) = 1 - alphas(i)
    }

    Factor.withVariables(phenotype, genotype).row(genotype, at(phenotype, 0), alphasCompliments).row(genotype, at(phenotype, 1), alphas).build()
  }

  private def getPhenotypeGivenGenotypeMendelianFactor(isDominant: Boolean, genotype: Var, phenotype: Var): Factor = {
    assert(genotype.cardinality == 3)
    assert(phenotype.cardinality == 2)
    if (isDominant) {
      Factor.withVariables(phenotype, genotype)
        .row(phenotype, at(genotype, "FF"), List("T", "F"), Array[Double](1, 0))
        .row(phenotype, at(genotype, "Ff"), List("T", "F"), Array[Double](1, 0))
        .row(phenotype, at(genotype, "ff"), List("T", "F"), Array[Double](0, 1))
        .build()
    }
    else {
      Factor.withVariables(phenotype, genotype)
        .row(phenotype, at(genotype, "FF"), List("T", "F"), Array[Double](0, 1))
        .row(phenotype, at(genotype, "Ff"), List("T", "F"), Array[Double](0, 1))
        .row(phenotype, at(genotype, "ff"), List("T", "F"), Array[Double](1, 0))
        .build()
    }
  }
}